homeassistant/components/trafikverket_train/sensor.py

```diff
@@ -0,0 +1,177 @@
+"""Train information for departures and delays, provided by Trafikverket."""
+
+from datetime import date, datetime, timedelta
+import logging
+
+from pytrafikverket import TrafikverketTrain
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import CONF_API_KEY, CONF_NAME, CONF_WEEKDAY, WEEKDAYS
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_TRAINS = "trains"
+CONF_FROM = "from"
+CONF_TO = "to"
+CONF_TIME = "time"
+
+ATTR_CANCELED = "canceled"
+ATTR_DELAY_TIME = "number_of_minutes_delayed"
+ATTR_PLANNED_TIME = "planned_time"
+ATTR_ESTIMATED_TIME = "estimated_time"
+ATTR_ACTUAL_TIME = "actual_time"
+ATTR_OTHER_INFORMATION = "other_information"
+ATTR_DEVIATIONS = "deviations"
+
+ICON = "mdi:train"
+SCAN_INTERVAL = timedelta(minutes=5)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Required(CONF_TRAINS): [{
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_TO): cv.string,
+        vol.Required(CONF_FROM): cv.string,
+        vol.Optional(CONF_TIME): cv.time,
+        vol.Optional(CONF_WEEKDAY, default=WEEKDAYS):
+            vol.All(cv.ensure_list, [vol.In(WEEKDAYS)])}]
+})
+
+
+async def async_setup_platform(
+        hass, config, async_add_entities, discovery_info=None):
+    """Set up the departure sensor."""
+    httpsession = async_get_clientsession(hass)
+    train_api = TrafikverketTrain(httpsession, config[CONF_API_KEY])
+    sensors = []
+    station_cache = {}
+    for train in config[CONF_TRAINS]:
+        try:
+            trainstops = [train[CONF_FROM], train[CONF_TO]]
+            for station in trainstops:
+                if station not in station_cache:
+                    station_cache[station] = await \
+                        train_api.async_get_train_station(station)
+
+        except ValueError as station_error:
+            station_error = str(station_error)
```

Comment:
We don't need to copy to string. The exception argument has a `__str__` magic method so we can use it as a logging message formatting argument directly.

Reply:
Hmm, the reason i forced it to a string was that the lookup of text in the exception message fail otherwise. 

Maybe there are other possibilities?

Reply:
This one works for looking up content in error message: 
`if "Invalid authentication" in station_error.args[0]:`

Reply:
Pushed change